javascript/nodejsscan/archive_path_overwrite.yaml

```diff
@@ -0,0 +1,154 @@
+      cwe: >-
+        CWE-22: Improper Limitation of a Pathname to a Restricted Directory
+        ('Path Traversal')
+  - id: zip_path_overwrite2
+    patterns:
+      - pattern-either:
+          - pattern-inside: |
+              var $X = require('unzip');
+              ...
+          - pattern-inside: |
+              var $X = require('unzipper');
+              ...
+      - pattern-inside: |
+          $UNZIP.Parse(...).on('entry', function $FUNC($ENTRY) {
+              ...
+          }, ...);
+      - pattern-not: |
+          if ($FILENAME.indexOf('..'));
+      - pattern-not: >
+          $FS.createWriteStream($PATH.join(..., $PATH.basename($FILENAME,
+          ...)));
+      - pattern-not: |
+          $FS.writeFile($PATH.join(..., $PATH.basename($FILENAME, ...)));
+      - pattern-not: |
+          $FS.writeFileSync($PATH.join(..., $PATH.basename($FILENAME, ...)));
+      - pattern-either:
+          - pattern: |
+              $FS.createWriteStream($FIL, ...);
+          - pattern: |
+              $FS.writeFile($FIL, ...);
+          - pattern: |
+              $FS.writeFileSync($FIL, ...);
+    message: >-
+      Insecure ZIP archive extraction can result in arbitrary path over write
+      and can result in code injection.
+    languages:
+      - js
+    severity: ERROR
+    metadata:
+      owasp: 'A5: Broken Access Control'
+      cwe: >-
+        CWE-22: Improper Limitation of a Pathname to a Restricted Directory
+        ('Path Traversal')
+  - id: admzip_path_overwrite
+    patterns:
+      - pattern-inside: |
+          var $X = require('adm-zip');
+          ...
+      - pattern-not: |
+          if ($FILENAME.indexOf('..'));
+      - pattern-not: >
+          $FS.createWriteStream($PATH.join(..., $PATH.basename($FILENAME,
+          ...)));
+      - pattern-not: |
+          $FS.writeFile($PATH.join(..., $PATH.basename($FILENAME, ...)));
+      - pattern-not: |
+          $FS.writeFileSync($PATH.join(..., $PATH.basename($FILENAME, ...)));
+      - pattern-either:
+          - pattern: >-
+              $ZIPENTZ.forEach(function $FUNC($ENTRY, ...) {
+              $FS.createWriteStream(...); }, ...);
+          - pattern: >-
+              $ZIPENTZ.forEach(function $FUNC($ENTRY, ...) { $FS.writeFile(...);
+              }, ...);
+          - pattern: >-
+              $ZIPENTZ.forEach(function $FUNC($ENTRY, ...) {
+              $FS.writeFileSync(...); }, ...);
+    message: >-
+      Insecure ZIP archive extraction using adm-zip can result in arbitrary path
+      over write and can result in code injection.
+    languages:
+      - js
+    severity: ERROR
+    metadata:
+      owasp: 'A5: Broken Access Control'
+      cwe: >-
+        CWE-22: Improper Limitation of a Pathname to a Restricted Directory
+        ('Path Traversal')
+  - id: tar_path_overwrite
+    patterns:
+      - pattern-inside: |
+          var $X = require('tar-stream');
+          ...
+      - pattern-inside: |
```

Comment:
Regarding https://github.com/returntocorp/semgrep/issues/738:
One thing we could do to make tests pass is add

```
      - pattern-not-inside: |
          $Y.pipe($UNZIP.Parse(...)).on('entry', function $FUNC(...) {
              ...
          }, ...);
```
here.  `tar_path_overwrite` is triggering on the `unzipper` test at the bottom of the file because it meets the requirements for 

```
- pattern-inside: |
          $EXTRACT.on('entry', function $FUNC(...) {
            ...
          }, ...);
```
because `$EXTRACT` can include an expression that matches `$SOMETHING.Parse(...)`.

However, the easier thing to do is just move the unzipper test up in the file. :)

In order to keep moving with these, I'm gonna move the test. If changes need to be made later, we can always follow up with new PRs. :)

Reply:
Cool. I fixed that.